geoutils/georaster/raster.py

```diff
@@ -958,12 +1002,101 @@ def copy(self: RasterType, new_array: np.ndarray | None = None) -> RasterType:
+        method: str,
+        *inputs: Raster | tuple[Raster, Raster],
+        **kwargs: Any,
+    ) -> Raster | tuple[Raster, Raster]:
+        """
+        Method to cast NumPy universal functions directly on Raster classes, by passing to the masked array.
+        This function basically applies the ufunc (with its method and kwargs) to .data, and rebuilds the Raster from
+        self.__class__. The cases separate the number of input nin and output nout, to properly feed .data and return
+        Raster objects.
+        See more details in NumPy doc, e.g., https://numpy.org/doc/stable/user/basics.dispatch.html#basics-dispatch.
+        """
+
+        # If the universal function takes only one input
+        if ufunc.nin == 1:
+            # If the universal function has only one output
+            if ufunc.nout == 1:
+                return self.__class__(
+                    {
+                        "data": getattr(ufunc, method)(inputs[0].data, **kwargs),  # type: ignore
+                        "transform": self.transform,
+                        "crs": self.crs,
+                        "nodata": self.nodata,
+                    }
+                )
+
+            # If the universal function has two outputs (Note: no ufunc exists that has three outputs or more)
+            else:
+                output = getattr(ufunc, method)(inputs[0].data, **kwargs)  # type: ignore
+                return self.__class__(
+                    {"data": output[0], "transform": self.transform, "crs": self.crs, "nodata": self.nodata}
+                ), self.__class__(
+                    {"data": output[1], "transform": self.transform, "crs": self.crs, "nodata": self.nodata}
+                )
+
+        # If the universal function takes two inputs (Note: no ufunc exists that has three inputs or more)
+        else:
+            if ufunc.nout == 1:
+                return self.__class__(
+                    {
+                        "data": getattr(ufunc, method)(inputs[0].data, inputs[1].data, **kwargs),  # type: ignore
+                        "transform": self.transform,
+                        "crs": self.crs,
+                        "nodata": self.nodata,
+                    }
+                )
+
+            # If the universal function has two outputs (Note: no ufunc exists that has three outputs or more)
+            else:
+                output = getattr(ufunc, method)(inputs[0].data, inputs[1].data, **kwargs)  # type: ignore
+                return self.__class__(
+                    {"data": output[0], "transform": self.transform, "crs": self.crs, "nodata": self.nodata}
+                ), self.__class__(
+                    {"data": output[1], "transform": self.transform, "crs": self.crs, "nodata": self.nodata}
+                )
+
+    def __array_function__(self, func: Callable[[np.ndarray, Any], Any], types: type, args: Any, kwargs: Any) -> Any:
```

Comment:
what's the role/meaning of `types` here?

Reply:
It is passing the types of the input(s) on which the NumPy function was used, to ensure those apply to our class/subclass (potentially needs subclassing). More info in the NumPy doc.

Reply:
(and I changed the typing as it should have been: `tuple[type]`)